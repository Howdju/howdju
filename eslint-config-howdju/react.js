module.exports = {
  parserOptions: {
    "ecmaFeatures": {
      "jsx": true
    },
  },
  env: {
    "browser": true,
  },
  extends: [
    "plugin:react/recommended"
  ],
  plugins: [
    "react",
  ],
  rules: {
    indent: [
      "warn",
      2,
      {
        "SwitchCase": 1,
        "ignoredNodes": [
          // We indent to the indent of the first prop, which this rule doesn't support
          "JSXAttribute",
          "JSXSpreadAttribute",
        ],
      }
    ],
    "react/boolean-prop-naming": "error",
    "react/no-children-prop": [
      // react-md requires us to use a children prop
      "off",
    ],
    "react/prop-types": [
      // This rule can't pick up props added by redux's mapStateToProps/mapDispatchToProps
      "off",
    ],
    // This doesn't seem to help; need to disable indent on JSXAttribute
    // "react/jsx-indent-props": [
    //   // We indent to the indent of the first prop, which this rule doesn't support
    //   "off"
    // ],
    "require-yield": [
      // Some of our sagas are side-effect only, requiring no yields.  And I don't think there's any harm in this.
      "off"
    ],
  },
  settings: {
    react: {
      version: "detect",
    },
  },
};
