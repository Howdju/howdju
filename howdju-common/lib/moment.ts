import moment, { Moment } from "moment";

// @types/moment doesn't provide this constructor, but it works.
export type MomentConstructor = {
  new (...args: Parameters<typeof moment>): Moment;
};
export const MomentConstructor = moment as unknown as MomentConstructor;
