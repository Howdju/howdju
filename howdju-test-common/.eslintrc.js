module.exports = {
  overrides: [
    {
      files: ["**/*.{js,ts}"],
      excludedFiles: ["node_modules/**"],
      extends: ["howdju/node"],
    },
  ],
};
