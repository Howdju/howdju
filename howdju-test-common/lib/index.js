const sinon = require('sinon')

module.exports.mockLogger = {
  log: sinon.fake(console.log),
  exception: sinon.fake(console.error),
  error: sinon.fake(console.error),
  warning: sinon.fake(console.warn),
  info: sinon.fake(console.info),
  debug: sinon.fake(console.debug),
}
