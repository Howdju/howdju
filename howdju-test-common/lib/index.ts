/// <reference path="../../howdju-common/lib/dom-anchor-text-position.d.ts" />
/// <reference path="../../howdju-common/lib/dom-anchor-text-quote.d.ts" />

export * from "./domTestUtils";
export * from "./TestTopicMessageSender";
export * from "./mockLogger";
export * from "./testUtil";
