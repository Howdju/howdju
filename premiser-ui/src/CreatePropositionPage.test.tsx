import React from "react";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import { rest } from "msw";
import { merge } from "lodash";
import moment from "moment";
import { pathToRegexp } from "path-to-regexp";

import {
  brandedParse,
  CreateJustification,
  CreateProposition,
  httpStatusCodes,
  JustificationOut,
  PropositionOut,
  PropositionRef,
} from "howdju-common";
import {
  InferRequestBody,
  InferResponseBody,
  serviceRoutes,
} from "howdju-service-routes";
import { expectToBeSameMomentDeep } from "howdju-test-common";

import {
  ariaVisibleOne,
  makeRouteComponentProps,
  progressToBeGone,
  renderWithProviders,
  setupUserEvent,
  withFakeTimers,
  withMockServer,
} from "@/testUtils";
import CreatePropositionPage from "./CreatePropositionPage";

const server = withMockServer();

describe("CreatePropositionPage", () => {
  describe("(with fake timers)", () => {
    withFakeTimers();

    test("renders correctly with query params", () => {
      // Arrange
      const description = "A credible source";
      const quoteText = "An important conclusion.";
      const url = "https://www.info.com";

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps("submit", {
        searchParams: {
          description,
          quoteText,
          url,
        },
      });

      // Act
      const { container } = renderWithProviders(
        <CreatePropositionPage
          mode={"SUBMIT_JUSTIFICATION_VIA_QUERY_STRING"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );

      // Assert
      expect(
        ariaVisibleOne(screen.getAllByDisplayValue(description))
      ).toBeInTheDocument();
      expect(
        ariaVisibleOne(screen.getAllByDisplayValue(quoteText))
      ).toBeInTheDocument();
      expect(
        ariaVisibleOne(screen.getAllByDisplayValue(url))
      ).toBeInTheDocument();

      jest.runAllTimers();
      expect(container).toMatchSnapshot();
    });

    test("can add URL while editing a WritQuote-based justification", async () => {
      // Arrange
      const user = setupUserEvent();

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps("submit");

      const { container } = renderWithProviders(
        <CreatePropositionPage
          mode={"CREATE_PROPOSITION"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );
      await user.click(
        screen.getByRole("checkbox", { name: /create justification/i })
      );
      await user.click(
        screen.getByRole("radio", { name: /An external reference/i })
      );

      // Act
      await user.click(screen.getByRole("button", { name: /add url/i }));

      // Assert
      expect(await screen.findAllByLabelText(/url/)).toHaveLength(2);

      jest.runAllTimers();
      expect(container).toMatchSnapshot();
    });

    test("can submit a proposition justified via WritQuote query params", async () => {
      // Arrange
      const user = setupUserEvent();

      const description = "A credible source";
      const quoteText = "An important conclusion.";
      const url = "https://www.info.com";

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps("submit", {
        searchParams: {
          description,
          quoteText,
          url,
        },
      });

      renderWithProviders(
        <CreatePropositionPage
          mode={"SUBMIT_JUSTIFICATION_VIA_QUERY_STRING"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );

      const proposition: CreateProposition = {
        text: "A bonny wee proposition.",
      };
      const justification: CreateJustification = {
        target: {
          type: "PROPOSITION",
          entity: proposition,
        },
        polarity: "POSITIVE",
        basis: {
          type: "WRIT_QUOTE",
          entity: {
            quoteText,
            writ: {
              title: description,
            },
            urls: [{ url }],
          },
        },
      };

      const response: InferResponseBody<
        typeof serviceRoutes.createJustification
      > = {
        isExtant: false,
        justification: merge({}, justification, {
          id: "1582",
          target: { entity: { id: "9483" } },
          basis: { entity: { id: "910" } },
          rootTargetType: "PROPOSITION",
          rootTarget: { id: "9483" },
        }) as JustificationOut,
      };
      let requestBody;
      server.use(
        rest.post(`http://localhost/justifications`, async (req, res, ctx) => {
          requestBody = await req.json();
          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }),
        rest.get("http://localhost/search-propositions", (_req, res, ctx) => {
          // delay so that we can cancel the request.
          return res(ctx.delay(1), ctx.json([]));
        })
      );
      // Use click/keyboard instead of type to trigger autocompletes.
      await user.click(screen.getByLabelText(/text/i));
      for (const letter of proposition.text) {
        await user.keyboard(letter);
        // Run timers in between keystrokes to allow sagas to run?
        jest.runAllTimers();
      }

      // Act
      await user.click(screen.getByRole("button", { name: /create/i }));

      // Assert
      jest.runAllTimers();
      expect(requestBody).toMatchObject({
        justification,
      });
      // TODO(196): get path pattern from routesById instead.
      expect(history.location.pathname).toMatch(pathToRegexp("/p/:id"));
    });

    test("can submit a proposition justified via a new WritQote", async () => {
      // Arrange
      const user = setupUserEvent();

      const quoteText = "An important conclusion.";
      const title = "A credible source";
      const url = "https://www.info.com";

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps("create-proposition");

      renderWithProviders(
        <CreatePropositionPage
          mode={"CREATE_PROPOSITION"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );

      const proposition: CreateProposition = {
        text: "A bonny wee proposition.",
      };
      const justification: CreateJustification = {
        target: {
          type: "PROPOSITION",
          entity: proposition,
        },
        polarity: "POSITIVE",
        basis: {
          type: "WRIT_QUOTE",
          entity: {
            quoteText,
            writ: {
              title,
            },
            urls: [{ url }],
          },
        },
      };

      const response: InferResponseBody<
        typeof serviceRoutes.createJustification
      > = {
        isExtant: false,
        justification: merge({}, justification, {
          id: "1582",
          target: { entity: { id: "9483" } },
          basis: { entity: { id: "910" } },
          rootTargetType: "PROPOSITION",
          rootTarget: { id: "9483" },
        }) as JustificationOut,
      };
      let requestBody;
      server.use(
        rest.post(`http://localhost/justifications`, async (req, res, ctx) => {
          requestBody = await req.json();
          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }),
        rest.get("http://localhost/search-propositions", (_req, res, ctx) => {
          // delay so that we can cancel the request.
          return res(ctx.delay(1), ctx.json([]));
        }),
        rest.get("http://localhost/search-writs", (_req, res, ctx) => {
          // delay so that we can cancel the request.
          return res(ctx.delay(1), ctx.json([]));
        })
      );

      await user.type(
        document.querySelector('textarea[name="proposition.text"]')!,
        proposition.text
      );
      const writQuoteRadio = Array.from(
        document.querySelectorAll('input[name="justification.basis.type"]')
      ).find((el) => /Quote/i.test(el.parentElement?.textContent!));
      await user.click(writQuoteRadio!);
      await user.type(
        document.querySelectorAll(".writ-quote-editor-fields textarea")[0],
        quoteText
      );
      await user.type(screen.getByLabelText(/Title/i), title);
      await user.type(screen.getByLabelText(/URL/i), url);

      // Act
      await user.click(screen.getByRole("button", { name: /create/i }));

      // Assert
      jest.runAllTimers();
      expect(requestBody).toMatchObject({
        justification,
      });
      // TODO(196): get path pattern from routesById instead.
      expect(history.location.pathname).toMatch(pathToRegexp("/p/:id"));
    });

    test("removing a tag removes it", async () => {
      // Arrange
      const user = setupUserEvent();

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps("submit");

      renderWithProviders(
        <CreatePropositionPage
          mode={"CREATE_PROPOSITION"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );

      const tagName = "TestTag";
      await user.type(screen.getByLabelText(/tag/i), tagName);
      await user.type(screen.getByLabelText(/tag/i), "{Enter}");

      // Act
      await user.click(document.querySelector(".remove-chip-icon") as Element);

      // Assert
      jest.runAllTimers();
      expect(
        screen.queryByRole("button", { name: new RegExp(tagName) })
      ).not.toBeInTheDocument();
    });

    test("attempting to submit with an invalid form shows errors", async () => {
      const user = setupUserEvent();

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps("submit");

      renderWithProviders(
        <CreatePropositionPage
          mode={"CREATE_PROPOSITION"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );

      const tagName = "TestTag";
      await user.type(screen.getByLabelText(/tag/i), tagName);
      await user.type(screen.getByLabelText(/tag/i), "{Enter}");

      // Act
      await user.click(screen.getByRole("button", { name: /create/i }));

      // Assert
      screen
        .getAllByText("String must contain at least 1 character(s)")
        .forEach((el) => expect(el).toBeInTheDocument());
    });
  });

  describe("(with real timers)", () => {
    test("can submit a Proposition justified by an extant Proposition", async () => {
      // I'm not sure why we need real timers for this test, but what I observe is that
      // fetchAndBeginEditOfNewJustificationFromBasisSource puts the beginEdit, but then the
      // component never re-renders based upon the updated state. I even checked that the reducer
      // was returning an unequal state.

      // Arrange
      const user = userEvent.setup();

      const basisSourceId = "2";
      const basisSourceType = "PROPOSITION";

      const history = createMemoryHistory();
      const { location, match } = makeRouteComponentProps(
        "create-justification",
        {
          searchParams: {
            basisSourceId,
            basisSourceType,
          },
        }
      );

      const proposition: CreateProposition = {
        text: "A bonny wee proposition.",
      };
      const basisProposition: PropositionOut = brandedParse(PropositionRef, {
        id: basisSourceId,
        text: "This warrants a response.",
        created: moment(),
      });
      const justification: CreateJustification = {
        target: {
          type: "PROPOSITION",
          entity: proposition,
        },
        polarity: "POSITIVE",
        basis: {
          type: "PROPOSITION_COMPOUND",
          entity: {
            atoms: [{ entity: basisProposition }],
          },
        },
      };

      let requestBody:
        | InferRequestBody<typeof serviceRoutes.createJustification>
        | undefined = undefined;
      server.use(
        rest.post(`http://localhost/justifications`, async (req, res, ctx) => {
          requestBody = await req.json();

          const response: InferResponseBody<
            typeof serviceRoutes.createJustification
          > = {
            isExtant: false,
            justification: merge({}, justification, {
              id: "1582",
              target: { entity: { id: "9483" } },
              basis: { entity: { id: "910" } },
              rootTargetType: "PROPOSITION",
              rootTarget: { id: "9483" },
            }) as JustificationOut,
          };

          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }),
        rest.get("http://localhost/search-propositions", (_req, res, ctx) => {
          // delay so that we can cancel the request.
          return res(ctx.delay(1), ctx.json([]));
        }),
        rest.get(
          `http://localhost/propositions/${basisProposition.id}`,
          (_req, res, ctx) => {
            const response: InferResponseBody<
              typeof serviceRoutes.readProposition
            > = {
              proposition: basisProposition,
            };
            // delay so that we can cancel the request.
            return res(ctx.delay(1), ctx.json(response));
          }
        )
      );

      renderWithProviders(
        <CreatePropositionPage
          mode={"CREATE_JUSTIFICATION"}
          history={history}
          location={location}
          match={match}
        />,
        { history }
      );

      await waitFor(() => {
        progressToBeGone(screen.queryByRole("progressbar"));
      });

      // Use click/keyboard instead of type to trigger autocompletes.
      const targetInput = screen.getAllByLabelText(/text/i)[0];
      await user.click(targetInput);
      for (const letter of proposition.text) {
        await user.keyboard(letter);
      }

      const createButton = screen.getByRole("button", { name: /create/i });
      await waitFor(() =>
        expect(createButton).not.toHaveClass("md-text--disabled")
      );

      // Act
      await user.click(createButton);

      // Assert
      await waitFor(() => expect(requestBody).toBeTruthy());
      expect(requestBody).toMatchObject(
        expectToBeSameMomentDeep({
          justification,
        })
      );
      // TODO(196): get path pattern from routesById instead.
      expect(history.location.pathname).toMatch(pathToRegexp("/p/:id"));
    });
  });
});
