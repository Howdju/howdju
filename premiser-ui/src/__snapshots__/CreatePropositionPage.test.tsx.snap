// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePropositionPage (with fake timers) renders correctly with query params 1`] = `
<div>
  <div
    id="edit-proposition-justification-page"
  >
    <form>
      <div
        class="md-grid"
      >
        <div
          class="rmd-card rmd-card--shadowed md-cell md-cell--12"
        >
          <header
            class="rmd-card__header rmd-card__header--center"
          >
            <span
              class="rmd-card__header-content"
            >
              <h5
                class="rmd-card__title"
              >
                Justify proposition
              </h5>
            </span>
          </header>
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            <button
              class="rmd-button rmd-button--text rmd-button--text-primary"
              title="Add Speaker"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rmd-icon rmd-icon--font material-icons rmd-icon--before"
              >
                person_add
              </i>
              Add Speaker
              <span
                class="rmd-ripple-container"
              />
            </button>
          </div>
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            <div>
              <ul
                class="error-message"
              />
              <div
                style="display: flex; align-items: center;"
              >
                <div
                  class="rmd-field-message-container"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-owns="create-proposition-page--proposition--text-listbox"
                    class="rmd-text-field-container rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--underline rmd-text-field-container--underline-labelled rmd-text-field-container--underline-active rmd-text-field-container--underline-left rmd-textarea-container rmd-textarea-container--animate rmd-textarea-container--cursor rmd-autocomplate api-autocomplete"
                    id="create-proposition-page--proposition--text-combobox"
                    role="combobox"
                    style="flex-grow: 1;"
                  >
                    <label
                      class="rmd-label rmd-label--active rmd-floating-label rmd-floating-label--active"
                      for="create-proposition-page--proposition--text"
                    >
                      Text
                    </label>
                    <div
                      class="rmd-textarea-container__inner rmd-textarea-container__inner--animate"
                    >
                      <textarea
                        aria-activedescendant=""
                        aria-autocomplete="list"
                        aria-controls="create-proposition-page--proposition--text-combobox"
                        autocomplete="off"
                        class="rmd-textarea rmd-textarea--floating rmd-textarea--rn"
                        id="create-proposition-page--proposition--text"
                        maxlength="512"
                        name="proposition.text"
                        required=""
                        rows="1"
                      />
                      <textarea
                        aria-hidden="true"
                        class="rmd-textarea rmd-textarea--rn rmd-textarea--mask rmd-textarea--floating"
                        id="create-proposition-page--proposition--text-mask"
                        readonly=""
                        rows="1"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                  <div
                    aria-live="polite"
                    class="rmd-form-message rmd-form-message--outline"
                    id="create-proposition-page--proposition--text-message"
                  >
                    <span
                      class="rmd-form-message__counter"
                      id="create-proposition-page--proposition--text-message-counter"
                    >
                      0 / 512
                    </span>
                  </div>
                </div>
                <input
                  style="display: none;"
                  type="submit"
                />
              </div>
            </div>
            <div
              class="chips-list"
            >
              <div
                style="display: flex; align-items: center;"
              >
                <div
                  class="rmd-field-message-container"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-owns="create-proposition-page--tags--tag-name-listbox"
                    class="rmd-text-field-container rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--underline rmd-text-field-container--underline-labelled rmd-text-field-container--underline-left rmd-textarea-container rmd-textarea-container--animate rmd-textarea-container--cursor rmd-autocomplate api-autocomplete"
                    id="create-proposition-page--tags--tag-name-combobox"
                    role="combobox"
                    style="flex-grow: 1;"
                  >
                    <label
                      class="rmd-label rmd-floating-label"
                      for="create-proposition-page--tags--tag-name"
                    >
                      Tag
                    </label>
                    <div
                      class="rmd-textarea-container__inner rmd-textarea-container__inner--animate"
                    >
                      <textarea
                        aria-activedescendant=""
                        aria-autocomplete="list"
                        aria-controls="create-proposition-page--tags--tag-name-combobox"
                        autocomplete="off"
                        class="rmd-textarea rmd-textarea--floating rmd-textarea--rn"
                        id="create-proposition-page--tags--tag-name"
                        maxlength="64"
                        name="tagName"
                        rows="1"
                      />
                      <textarea
                        aria-hidden="true"
                        class="rmd-textarea rmd-textarea--rn rmd-textarea--mask rmd-textarea--floating"
                        id="create-proposition-page--tags--tag-name-mask"
                        readonly=""
                        rows="1"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                  <div
                    aria-live="polite"
                    class="rmd-form-message rmd-form-message--outline"
                    id="create-proposition-page--tags--tag-name-message"
                  >
                    <span
                      class="rmd-form-message__counter"
                      id="create-proposition-page--tags--tag-name-message-counter"
                    >
                      0 / 64
                    </span>
                  </div>
                </div>
                <input
                  style="display: none;"
                  type="submit"
                />
              </div>
            </div>
          </div>
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            <div
              class="md-selection-control-container md-switch-container"
            >
              <input
                aria-hidden="true"
                checked=""
                class="md-selection-control-input"
                id="create-proposition-page--do-create-justification"
                name="doCreateJustification"
                type="checkbox"
                value=""
              />
              <label
                class="md-selection-control-label md-pointer--hover md-text"
                for="create-proposition-page--do-create-justification"
              >
                <div
                  class="md-switch-track md-pointer--hover md-switch-track--on"
                >
                  <div
                    aria-checked="true"
                    aria-labelledby="create-proposition-page--do-create-justification-label"
                    aria-pressed="false"
                    class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-switch-thumb md-switch-thumb--on"
                    role="checkbox"
                    tabindex="0"
                  >
                    <div
                      class="md-ink-container md-ink-container--2x"
                    />
                  </div>
                </div>
                <span
                  id="create-proposition-page--do-create-justification-label"
                >
                  Create justification
                </span>
              </label>
            </div>
          </div>
          <header
            class="rmd-card__header rmd-card__header--center"
          >
            <span
              class="rmd-card__header-content"
            >
              <h5
                class="rmd-card__title"
                title="Justification"
              />
            </span>
          </header>
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            <div>
              <div
                class=""
              >
                <fieldset
                  class="md-selection-control-group"
                >
                  <div
                    class="md-selection-control-container md-selection-control-container--inline"
                    title="Support the truth of the proposition"
                  >
                    <input
                      aria-hidden="true"
                      checked=""
                      class="md-selection-control-input"
                      id="create-proposition-page--justification--polarity0"
                      name="justification.polarity"
                      type="radio"
                      value="POSITIVE"
                    />
                    <label
                      class="md-selection-control-label md-pointer--hover md-text"
                      for="create-proposition-page--justification--polarity0"
                    >
                      <div
                        aria-checked="true"
                        aria-describedby="create-proposition-page--justification--polarity-group-label"
                        aria-labelledby="create-proposition-page--justification--polarity0-label"
                        aria-pressed="false"
                        class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-selection-control-toggle md-btn md-btn--icon md-text--theme-secondary"
                        role="radio"
                        tabindex="0"
                      >
                        <div
                          class="md-ink-container"
                        />
                        <i
                          class="md-icon material-icons md-text--inherit"
                        >
                          radio_button_checked
                        </i>
                      </div>
                      <span
                        id="create-proposition-page--justification--polarity0-label"
                      >
                        Supporting
                      </span>
                    </label>
                  </div>
                  <div
                    class="md-selection-control-container md-selection-control-container--inline"
                    title="Oppose the truth of the proposition"
                  >
                    <input
                      aria-hidden="true"
                      class="md-selection-control-input"
                      id="create-proposition-page--justification--polarity1"
                      name="justification.polarity"
                      type="radio"
                      value="NEGATIVE"
                    />
                    <label
                      class="md-selection-control-label md-pointer--hover md-text"
                      for="create-proposition-page--justification--polarity1"
                    >
                      <div
                        aria-checked="false"
                        aria-describedby="create-proposition-page--justification--polarity-group-label"
                        aria-labelledby="create-proposition-page--justification--polarity1-label"
                        aria-pressed="false"
                        class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-selection-control-toggle md-btn md-btn--icon md-text--secondary"
                        role="radio"
                        tabindex="-1"
                      >
                        <div
                          class="md-ink-container"
                        />
                        <i
                          class="md-icon material-icons md-text--inherit"
                        >
                          radio_button_unchecked
                        </i>
                      </div>
                      <span
                        id="create-proposition-page--justification--polarity1-label"
                      >
                        Opposing
                      </span>
                    </label>
                  </div>
                </fieldset>
              </div>
              <hr
                class="rmd-divider"
                role="separator"
              />
              <h6
                class="rmd-typography rmd-typography--subtitle-1"
              >
                Type
              </h6>
              <div
                class=""
              >
                <fieldset
                  class="md-selection-control-group"
                >
                  <div
                    class="md-selection-control-container md-selection-control-container--inline"
                  >
                    <input
                      aria-hidden="true"
                      class="md-selection-control-input"
                      id="create-proposition-page--justification--basis-type0"
                      name="justification.basis.type"
                      type="radio"
                      value="PROPOSITION_COMPOUND"
                    />
                    <label
                      class="md-selection-control-label md-pointer--hover md-text"
                      for="create-proposition-page--justification--basis-type0"
                    >
                      <div
                        aria-checked="false"
                        aria-describedby="create-proposition-page--justification--basis-type-group-label"
                        aria-labelledby="create-proposition-page--justification--basis-type0-label"
                        aria-pressed="false"
                        class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-selection-control-toggle md-btn md-btn--icon md-text--secondary"
                        role="radio"
                        tabindex="0"
                      >
                        <div
                          class="md-ink-container"
                        />
                        <i
                          class="md-icon material-icons md-text--inherit"
                        >
                          radio_button_unchecked
                        </i>
                      </div>
                      <span
                        id="create-proposition-page--justification--basis-type0-label"
                      >
                        <div
                          title="An argument based on an ordered list of propositions"
                        >
                          Proposition(s)
                        </div>
                      </span>
                    </label>
                  </div>
                  <div
                    class="md-selection-control-container md-selection-control-container--inline"
                  >
                    <input
                      aria-hidden="true"
                      class="md-selection-control-input"
                      id="create-proposition-page--justification--basis-type1"
                      name="justification.basis.type"
                      type="radio"
                      value="MEDIA_EXCERPT"
                    />
                    <label
                      class="md-selection-control-label md-pointer--hover md-text"
                      for="create-proposition-page--justification--basis-type1"
                    >
                      <div
                        aria-checked="false"
                        aria-describedby="create-proposition-page--justification--basis-type-group-label"
                        aria-labelledby="create-proposition-page--justification--basis-type1-label"
                        aria-pressed="false"
                        class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-selection-control-toggle md-btn md-btn--icon md-text--secondary"
                        role="radio"
                        tabindex="-1"
                      >
                        <div
                          class="md-ink-container"
                        />
                        <i
                          class="md-icon material-icons md-text--inherit"
                        >
                          radio_button_unchecked
                        </i>
                      </div>
                      <span
                        id="create-proposition-page--justification--basis-type1-label"
                      >
                        <div
                          title="Evidence based on an excerpt of an external media source"
                        >
                          Media Excerpt
                        </div>
                      </span>
                    </label>
                  </div>
                </fieldset>
              </div>
              <hr
                class="rmd-divider"
                role="separator"
              />
              <div
                class="writ-quote-editor-fields"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-textarea-container rmd-textarea-container--animate rmd-textarea-container--cursor"
                >
                  <label
                    class="rmd-label rmd-floating-label rmd-floating-label--active rmd-floating-label--inactive"
                    for="create-proposition-page--justification--basis-writ-quote--quote-text"
                  >
                    Quote
                  </label>
                  <div
                    class="rmd-textarea-container__inner rmd-textarea-container__inner--animate"
                  >
                    <textarea
                      class="rmd-textarea rmd-textarea--floating rmd-textarea--rn"
                      id="create-proposition-page--justification--basis-writ-quote--quote-text"
                      maxlength="4096"
                      name="justification.basis.writQuote.quoteText"
                      rows="2"
                    >
                      An important conclusion.
                    </textarea>
                    <textarea
                      aria-hidden="true"
                      class="rmd-textarea rmd-textarea--rn rmd-textarea--mask rmd-textarea--floating"
                      id="create-proposition-page--justification--basis-writ-quote--quote-text-mask"
                      readonly=""
                      rows="2"
                      tabindex="-1"
                    >
                      An important conclusion.
                    </textarea>
                  </div>
                </div>
                <div
                  class="writ-editor-fields"
                >
                  <div
                    style="display: flex; align-items: center;"
                  >
                    <div
                      class="rmd-field-message-container"
                    >
                      <div
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-owns="create-proposition-page--justification--basis-writ-quote--writ--title-listbox"
                        class="rmd-text-field-container rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--underline rmd-text-field-container--underline-labelled rmd-text-field-container--underline-left rmd-textarea-container rmd-textarea-container--animate rmd-textarea-container--cursor rmd-autocomplate api-autocomplete"
                        id="create-proposition-page--justification--basis-writ-quote--writ--title-combobox"
                        role="combobox"
                        style="flex-grow: 1;"
                      >
                        <label
                          class="rmd-label rmd-floating-label rmd-floating-label--active rmd-floating-label--inactive"
                          for="create-proposition-page--justification--basis-writ-quote--writ--title"
                        >
                          Title
                        </label>
                        <div
                          class="rmd-textarea-container__inner rmd-textarea-container__inner--animate"
                        >
                          <textarea
                            aria-activedescendant=""
                            aria-autocomplete="list"
                            aria-controls="create-proposition-page--justification--basis-writ-quote--writ--title-combobox"
                            autocomplete="off"
                            class="rmd-textarea rmd-textarea--floating rmd-textarea--rn"
                            id="create-proposition-page--justification--basis-writ-quote--writ--title"
                            maxlength="512"
                            minlength="1"
                            name="justification.basis.writQuote.writ.title"
                            required=""
                            rows="1"
                          >
                            A credible source
                          </textarea>
                          <textarea
                            aria-hidden="true"
                            class="rmd-textarea rmd-textarea--rn rmd-textarea--mask rmd-textarea--floating"
                            id="create-proposition-page--justification--basis-writ-quote--writ--title-mask"
                            readonly=""
                            rows="1"
                            tabindex="-1"
                          >
                            A credible source
                          </textarea>
                        </div>
                      </div>
                      <div
                        aria-live="polite"
                        class="rmd-form-message rmd-form-message--outline"
                        id="create-proposition-page--justification--basis-writ-quote--writ--title-message"
                      >
                        <span
                          class="rmd-form-message__counter"
                          id="create-proposition-page--justification--basis-writ-quote--writ--title-message-counter"
                        >
                          17 / 512
                        </span>
                      </div>
                    </div>
                    <input
                      style="display: none;"
                      type="submit"
                    />
                  </div>
                </div>
                <div
                  style="display: flex;"
                >
                  <div
                    class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--outline-left writ-quote-url-input"
                    style="flex-grow: 1;"
                  >
                    <span
                      class="rmd-text-field-addon rmd-text-field-addon--presentational"
                    >
                      <span
                        class="material-symbols"
                      >
                        link
                      </span>
                    </span>
                    <label
                      class="rmd-label rmd-floating-label rmd-floating-label--active rmd-floating-label--inactive"
                      for="create-proposition-page--justification--basis-writ-quote--urls-0--url"
                    >
                      URL
                    </label>
                    <input
                      aria-label="url"
                      class="rmd-text-field rmd-text-field--floating"
                      id="create-proposition-page--justification--basis-writ-quote--urls-0--url"
                      name="justification.basis.writQuote.urls[0].url"
                      type="url"
                      value="https://www.info.com"
                    />
                  </div>
                  <button
                    class="rmd-button rmd-button--icon rmd-button--text-primary"
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="rmd-icon rmd-icon--font material-icons"
                    >
                      delete
                    </i>
                    <span
                      class="rmd-ripple-container"
                    />
                  </button>
                </div>
                <button
                  class="rmd-button rmd-button--text rmd-button--text-primary add-button"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="rmd-icon rmd-icon--font material-icons rmd-icon--before"
                  >
                    add
                  </i>
                  Add URL
                  <span
                    class="rmd-ripple-container"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="rmd-card__actions"
          >
            <button
              class="rmd-button rmd-button--text rmd-button--text-primary"
              type="button"
            >
              Cancel
              <span
                class="rmd-ripple-container"
              />
            </button>
            <button
              aria-disabled="true"
              class="rmd-button rmd-button--text rmd-button--contained rmd-button--disabled"
              title="A proposition is only as good as its justification"
              type="submit"
            >
              Create
              <span
                class="rmd-ripple-container"
              />
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
